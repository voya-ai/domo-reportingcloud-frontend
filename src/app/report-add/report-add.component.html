<h1>Add report</h1>

<div>
  <mat-form-field class="example-full-width">
    <input matInput placeholder="Report name" [(ngModel)]="reportName">
  </mat-form-field>
</div>

<div>
  <app-report-template-autocomplete
    (onSelectionChange)="templateSelectionChanged($event)">
  </app-report-template-autocomplete>
</div>

<div>
  <app-report-dataset-autocomplete
    (onSelectionChange)="datasetSelectionChanged($event)">
  </app-report-dataset-autocomplete>
</div>

<div *ngIf="columnsLoading">
  <mat-spinner></mat-spinner>
</div>

<div
  *ngIf="headerColumns.length > 0">
  <app-dataset-column-select
    [(columns)]="headerColumns"
    title="Header columns">
  </app-dataset-column-select>
</div>

<div
  *ngIf="contentColumns.length > 0">
  <app-dataset-column-select
    [(columns)]="contentColumns"
    title="Content columns">
  </app-dataset-column-select>
</div>

<button
  type="button"
  [disabled]="headerColumns.length == 0 || !selectedTemplate || !reportName"
  class="btn btn-primary"
  (click)="save()">Save</button>
