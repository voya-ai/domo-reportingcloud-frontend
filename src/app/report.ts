export interface Report {
  name?: string;
  slug: string;
  dataSet?: string;
  templateName?: string;
  headerColumns?: string[];
  contentColumns?: string[];
}
