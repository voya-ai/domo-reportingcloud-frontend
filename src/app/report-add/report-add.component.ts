import { Component, OnInit } from '@angular/core';
import { FormControl } from '@angular/forms';
import { Observable } from 'rxjs/Observable';
import 'rxjs/add/operator/map';
import 'rxjs/add/operator/startWith';
import 'rxjs/add/operator/toPromise';
import 'rxjs/add/operator/debounceTime';
import 'rxjs/add/operator/distinctUntilChanged';
import 'rxjs/add/observable/fromPromise';

import { DatasetService } from 'app/dataset.service';
import { ReportService } from 'app/report.service';

import { DatasetColumn } from 'app/dataset-column';
import { Report } from 'app/report';

import { Router } from '@angular/router';

@Component({
  selector: 'app-report-add',
  templateUrl: './report-add.component.html',
  styleUrls: ['./report-add.component.css']
})

export class ReportAddComponent implements OnInit {
  reportName: string;
  headerColumns: DatasetColumn[] = [];
  contentColumns: DatasetColumn[] = [];

  selectedDataset: any;
  selectedTemplate: any;

  columnsLoading: boolean = false;

  constructor(private _datasetService: DatasetService,
    private _reportService: ReportService,
    private _router: Router)
  { }

  ngOnInit() {}

  datasetSelectionChanged(selection) {
    this.selectedDataset = selection;
    this.headerColumns = [];
    this.contentColumns = [];

    this.columnsLoading = true;

    this._datasetService.getColumns(selection.id)
      .toPromise()
      .then(function(columns : Array<DatasetColumn>) {
        this.columnsLoading = false;
        
        // deep copy columns
        this.headerColumns = columns.map(function(column) {
          return { ...column };
        });
        this.contentColumns = columns.map(function(column) {
          return { ...column };
        });


      }.bind(this));

  }

  templateSelectionChanged(selection) {
    this.selectedTemplate = selection;
  }

  save() {
    this._reportService.save({
      name: this.reportName,
      slug: "",
      dataSet: this.selectedDataset.id,
      templateName: this.selectedTemplate.templateName,
      headerColumns: this.headerColumns.filter(column => column.checked).map(column => column.name),
      contentColumns: this.contentColumns.filter(column => column.checked).map(column => column.name)
    }).toPromise()
    .then(function(report: Report) {
      this._router.navigateByUrl(`report/${report.slug}`);
    }.bind(this));
  }
}
