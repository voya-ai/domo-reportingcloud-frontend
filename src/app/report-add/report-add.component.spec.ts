import { async, ComponentFixture, TestBed } from '@angular/core/testing';

import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { ReportAddComponent } from './report-add.component';
import {
  MatInputModule,
  MatProgressSpinnerModule
} from '@angular/material';

import { NoopAnimationsModule } from '@angular/platform-browser/animations';
import { RouterTestingModule } from '@angular/router/testing';

import { ReportService } from 'app/report.service';
import { DatasetService } from 'app/dataset.service';

import 'rxjs/add/observable/of';

import { MockComponent } from 'src/testing/component-stubs';

class MockDatasetService {}
class MockReportService {}

describe('ReportAddComponent', () => {
  let component: ReportAddComponent;
  let fixture: ComponentFixture<ReportAddComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      providers: [
        { provide: DatasetService, useClass: MockDatasetService },
        { provide: ReportService, useClass: MockReportService }
      ],
      declarations: [
        ReportAddComponent,
        MockComponent({ selector: 'app-report-template-autocomplete'}),
        MockComponent({ selector: 'app-report-dataset-autocomplete' }),
        MockComponent({ selector: 'app-dataset-column-select', inputs: [ 'columns', 'title' ] })
      ],
      imports: [
        NoopAnimationsModule,
        MatInputModule,
        MatProgressSpinnerModule,
        FormsModule,
        ReactiveFormsModule,
        RouterTestingModule.withRoutes([])
      ]
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(ReportAddComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should match the snapshot', () => {
    expect(fixture).toMatchSnapshot();
  });
});
