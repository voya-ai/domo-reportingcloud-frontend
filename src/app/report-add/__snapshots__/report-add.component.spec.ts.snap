// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReportAddComponent should match the snapshot 1`] = `
<app-report-add
  _datasetService={[Function MockDatasetService]}
  _reportService={[Function MockReportService]}
  _router={[Function Router]}
  columnsLoading="false"
  contentColumns={[Function Array]}
  headerColumns={[Function Array]}
>
  <h1>
    Add report
  </h1>
  
  <div>
    
    
    <mat-form-field
      class="example-full-width mat-input-container mat-form-field ng-tns-c2-0 mat-form-field-type-mat-input mat-form-field-can-float mat-primary ng-untouched ng-pristine ng-valid"
    >
      <div
        class="mat-input-wrapper mat-form-field-wrapper"
      >
        <div
          class="mat-input-flex mat-form-field-flex"
        >
          
          <div
            class="mat-input-infix mat-form-field-infix"
          >
            
      
            <input
              aria-invalid="false"
              class="mat-input-element mat-form-field-autofill-control ng-untouched ng-pristine ng-valid"
              id="mat-input-1"
              matinput=""
              ng-reflect-placeholder="Report name"
              placeholder="Report name"
            />
            
    
            <span
              class="mat-input-placeholder-wrapper mat-form-field-placeholder-wrapper"
            >
              
              <label
                aria-owns="mat-input-1"
                class="mat-input-placeholder mat-form-field-placeholder ng-tns-c2-0 mat-empty mat-form-field-empty"
                for="mat-input-1"
              >
                Report name 
                
              </label>
            </span>
          </div>
          
        </div>
        <div
          class="mat-input-underline mat-form-field-underline"
        >
          <span
            class="mat-input-ripple mat-form-field-ripple"
          />
        </div>
        <div
          class="mat-input-subscript-wrapper mat-form-field-subscript-wrapper"
          ng-reflect-ng-switch="hint"
        >
          
          
          <div
            class="mat-input-hint-wrapper mat-form-field-hint-wrapper ng-tns-c2-0 ng-trigger ng-trigger-transitionMessages"
            style="opacity: 1;"
          >
            
            <div
              class="mat-input-hint-spacer mat-form-field-hint-spacer"
            />
          </div>
        </div>
      </div>
    </mat-form-field>
    
  
  </div>
  
  <div>
    
    
    <app-report-template-autocomplete />
    
  
  </div>
  
  <div>
    
    
    <app-report-dataset-autocomplete />
    
  
  </div>
  
  
  
  
  
  
  
  <button
    class="btn btn-primary"
    disabled=""
    type="button"
  >
    Save
  </button>
  
</app-report-add>
`;
